src/Interpreters/Access/InterpreterGrantQuery.cpp

```diff
@@ -330,6 +331,67 @@ namespace
             updateGrantedAccessRightsAndRolesTemplate(*role, elements_to_grant, elements_to_revoke, roles_to_grant, roles_to_revoke, admin_option);
     }
 
+    template <typename T>
+    void grantCurrentGrantsTemplate(
+        T & grantee,
+        const AccessRightsElements & elements_to_grant,
+        const AccessRightsElements & elements_to_revoke,
+        std::shared_ptr<const ContextAccess> current_user_access,
+        bool grant_option)
+    {
+        if (!elements_to_revoke.empty())
+            grantee.access.revoke(elements_to_revoke);
+
+        /// We need to collect all current user's grant and filter them by grant option.
+        AccessRightsElements current_user_grantable_elements;
+        auto available_grant_elements = current_user_access->getAccessRights()->getElements();
+        std::copy_if(
+            available_grant_elements.begin(),
+            available_grant_elements.end(),
+            std::back_inserter(current_user_grantable_elements),
+            [](AccessRightsElement x) { return x.grant_option || x.is_partial_revoke; });
+
+        AccessRights current_user_rights;
+        for (auto & element : current_user_grantable_elements)
+        {
+            if (element.is_partial_revoke)
+                current_user_rights.revoke(element);
+            else
+                current_user_rights.grant(element);
+        }
+
+        /// If elements_to_grant was not specified it will grant all available for user grants.
+        /// Otherwise, we will intersect available grants with given set.
+        if (elements_to_grant.empty())
+        {
+            if (!grant_option)
+                current_user_rights.revokeGrantOption(AccessType::ALL);
+
+            grantee.access.makeUnion(current_user_rights);
+        }
+        else
+        {
+            AccessRights new_rights(elements_to_grant);
+            new_rights.makeIntersection(current_user_rights);
```

Comment:
`grantCurrentGrantsTemplate()` can be called multiple times if you have a statement like

```
GRANT CURRENT GRANTS(SELECT ON *.*) TO A, B, C
```

And this intersection of the current grants and `SELECT ON *.*` must be the same for `A`, `B` and `C`.

Reply:
To solve this issue it's necessary to calculate the intersection with the current user's grants before `access_control.update(grantees, update_func)` is called.